Rest spec and documentation (#54664)
elastic/elasticsearch: x-pack/plugin/identity-provider/docs/en/rest-api/idp-saml-init.asciidoc

[role="xpack"]
[[idp-saml-init]]
=== Generate a SAML authentication response message
++++
<titleabbrev>Generate SAML response</titleabbrev>
++++
Generates a SAML Response message to be sent to a Service Provider

[[idp-saml-init-request]]
==== {api-request-title}

`POST /_idp/saml/init`

[[idp-saml-init-prereqs]]
==== {api-prereq-title}

* To use this API, you must have a role that grants the `cluster:admin/idp/saml/init` privilege.

[[idp-saml-init-desc]]
==== {api-description-title}

This API generates a SAML Response message that should be sent to a Service Provider as part of an
IDP initiated or SP initiated SAML Single Sign On. This API expects the caller to present
credentials for the user that the SAML Response will be created for as "Secondary Authentication"

> using the `es-secondary-authorization` HTTP Request header.

Reviewer: FYI, @lcawl is working on some generic docs for secondary authentication, so which we can link to when they're ready.

> The SAML response is returned as an XML String and the caller of the API is responsible to instruct

Reviewer: ```suggestion
The SAML response is returned as a String that contains an XML document. The caller of the API is responsible to instruct
```

the end user's browser to make an HTTP Post request to the Service Provider with the SAML response
Base64 encoded.

[[idp-saml-init-body]]
==== {api-request-body-title}

The following parameters can be specified in the body of a POST request:

`entity_id`::
(Required, string) The SAML entity Id of the service provider that will be the recipient of the SAML Response.

`acs`::
(Required, string) The assertion consumer service URL of the service provider that will be the recipient of the SAML Response.

`authn_state`::
(Optional, object) The JSON structure that <<idp-saml-validate>> returns. This should only be
provided when calling this API as part of an SP initiated Single Sign On.

[[idp-saml-init-example]]
==== {api-examples-title}

The following example generates a SAML Response for an IDP initiated SAML Single Sign On to the Service Provider with entity Id
`https://sp1.kibana.org` and an assertion consumer service URL that is `https://sp1.kibana.org/saml/acs`

[source, sh]
--------------------------------------------------------------------

> curl -u idp_admin:idp_admin_pwd <1> -H 'Content-Type: application/json' \

Reviewer: ```suggestion
curl -u idp_admin:idp_admin_pwd <1> \
  -H 'Content-Type: application/json' \
```
I think this will be neater if the footnotes are always at the end of line (but that's a personal view)

-H 'es-secondary-authorization: ApiKey dVhmUDBuQUJERWhZWEZaQVg5S0k6WUJubmZwNEtRZ1d4cGRxdXBzZmFDUQ==' <2> \
localhost:9200/_idp/saml/init -d '{"entity_id":"https://sp1.kibana.org","acs":"https://sp1.kibana.org/saml/acs"}'
--------------------------------------------------------------------
// NOTCONSOLE
<1> The credentials of the user that has the necessary privileges to call this API
<2> The credentials of the end user for which the SAML Response will be generated. These can be in the form of a Basic authentication

> header, an elasticsearch access token, or an API key.

Reviewer: ```suggestion
header, an {es} access token, or an API key.
```

  "saml_status" : "urn:oasis:names:tc:SAML:2.0:status:Requester",
  "error" : "User [user1] is not permitted to access service [https://sp1.kibana.org]",
  "service_provider" : {
    "entity_id" : "https://sp1.kibana.org"
  }
}
--------------------------------------------------------------------
// TESTRESPONSE[skip:Do not enable identity provider for the docs cluster, at least not yet]

A failed call, in the case of an IDP initiated SSO returns an error with the appropriate HTTP Status code. i.e. if the value of the

> `es-secondary-authorization` is wrong, the IDP will respond with `403`

Reviewer 2: From my experimentations, I got 403s when I tried to hit this endpoint with a user that doesn't have access to the resource, and 401s when the `es-secondary-authorization` is wrong.

Author: > I got 403s when I tried to hit this endpoint with a user that doesn't have access to the resource, and 401s when the es-secondary-authorization is wrong.

Yes, `403` is a typo in the original description, it should be `401`. `403` is normal/expected when hitting any ES endpoint with a user that is not authorized to do so , thus not sure we will explicitly call this out in this API docs.

> EDIT: and I also wasn't able to to get the 403 to show up when using basic auth..

Not sure what you mean here. Basic auth for `Authorization` header or for `es-secondary-authorization` ? And you were trying to get this for a reason, or because the description was wrong and you expected to get a `403` ?

Reviewer 2: Ah, if 403 is not supposed to be called out in this API, then we can leave it.

In any case, my issue was regarding discrepancy I saw: which is that when I was using basic `es-secondary-authorization`, I didn't get a 403 when using this endpoint to access a resource the user didn't have access to, but I did when I used ApiKey `es-secondary-authorization` auth.

Reviewer 2: Just following up: this was expected behaviour and had nothing to do with ApiKey vs Basic Auth. It was a per-design difference between Sp-initiated vs IdP initiated flow.
